pwndbg/aglib/kernel/vmmap.py

```diff
@@ -275,69 +345,35 @@ def kernel_vmmap_via_monitor_info_mem() -> Tuple[pwndbg.lib.memory.Page, ...]:
-        # these cases in a `finally` block instead of an `except` block.
-        # TODO: Find out which other architectures don't support this command
-        if pwndbg.aglib.arch.name == "aarch64":
-            print(
-                M.error(
-                    f"The {pwndbg.aglib.arch.name} architecture does"
-                    " not support the `monitor info mem` command.\n"
-                    "Run `help show kernel-vmmap` for other options."
-                )
-            )
-        return ()
+        # return a string starting with 'unknown command:'.
+        monitor_info_mem = "unknown command"
 
-    lines = monitor_info_mem.splitlines()
+    parser_func = None
+    if pwndbg.aglib.arch.name in ("i386", "x86-64"):
+        parser_func = _parser_mem_info_line_x86
+    elif pwndbg.aglib.arch.name == "rv64":
+        parser_func = _parser_mem_info_line_riscv64
 
-    # Handle disabled PG
-    # This will prevent a crash on abstract architectures
-    if len(lines) == 1 and lines[0] == "PG disabled":
+    if parser_func is None or "unknown command" in monitor_info_mem:
+        print(
+            M.error(
+                f"The {pwndbg.aglib.arch.name} architecture does"
+                " not support the `monitor info mem` command.\n"
+                "Run `help show kernel-vmmap` for other options."
+            )
+        )
         return ()
 
-    global monitor_info_mem_not_warned
     pages: List[pwndbg.lib.memory.Page] = []
-    for line in lines:
+    for line in monitor_info_mem.splitlines():
         try:
-            dash_idx = line.index("-")
-            space_idx = line.index(" ")
-            rspace_idx = line.rindex(" ")
-
-            start = int(line[:dash_idx], 16)
-            end = int(line[dash_idx + 1 : space_idx], 16)
-            size = int(line[space_idx + 1 : rspace_idx], 16)
+            page = parser_func(line)
         except Exception:
             # invalid format
             continue
-        if end - start != size and monitor_info_mem_not_warned:
-            print(
-                M.warn(
-                    (
-                        "The vmmap output may be incorrect as `monitor info mem` output assertion/assumption\n"
-                        "that end-start==size failed. The values are:\n"
-                        "end=%#x; start=%#x; size=%#x; end-start=%#x\n"
-                        "Note that this warning will not show up again in this Pwndbg/GDB session."
-                    )
-                    % (end, start, size, end - start)
-                )
-            )
-            monitor_info_mem_not_warned = False
-        perm = line[rspace_idx + 1 :]
-
-        flags = 0
-        if "r" in perm:
-            flags |= 4
-        if "w" in perm:
-            flags |= 2
-        if "x" in perm:
-            flags |= 1
-        pages.append(pwndbg.lib.memory.Page(start, size, flags, 0, "<qemu>"))
+        pages.append(page)
```

Comment:
Ideally, this should be: `pages = [parser_func(line) for line in monitor_info_mem.splitlines()]`